### PR TITLE
Add new string subcommands

### PR DESCRIPTION
**Pull request by dbohdan**
In this PR I have implemented the following `string` subcommands:
* `string tolower`, `string toupper`
* `string first`, `string last`
* `string trim`, `string trimleft`, `string trimright`
* `string map`
* `string range`

I have added over 60 tests for them in total.  I didn't make any changes to the Molt Book.  Should contributors do that?  The implementation of `string map` seems a little ugly.  I am not sure how to improve it.

Resolved issues: #75, #77, #79, #80, #84, #85, #86.

### CODE REVIEW

**Comment by wduquette** [comment]
What a pleasant surprise!

I'll look through the code first thing tomorrow morning, and let you know if I have any heartburn; otherwise I'll go ahead and merge it.

Regarding the Molt Book, you're welcome to take a whack at it; the big deal, as I'm sure you've noticed, is to:

* Describe the implemented semantics and syntax clearly.
* Catalog any "TCL Liens" for each subcommand.
* Follow the existing format (including updating the link table at the top).

You already know the "TCL Liens" for the new implementation, if any; if you update the Molt Book `string.md` page, I might change the wording a bit but it makes me less likely to miss something.  Similar for the rustdocs for the public API.

The "how to" sections of the Molt Book I probably want to update myself.

Thank you!

**Comment by wduquette** [comment]
Thank you especially for providing `string map`, which I'd been dreading.

**Comment by dbohdan** [comment]
You're welcome!  I look forward to your feedback.  As for the Tcl Liens, I see two that are relevant to the subcommands I have implemented but not specific to them or `string`.

1. Tcl 8 and Rust handle Unicode differently.  I haven't compared them closely.  Molt must at least lack the `TCL_UTF_MAX` limitations that standard builds of Tcl 8.6 have.  The particular differences need to be investigated more substantially.
2. `string range` does not understand the `integer[+-]integer` or the `end[+-]integer` syntax for indices.  There is already a note about this limitation on the Molt Book page for `lindex`.  Since the limitation applies to all range and index commands, I think this note should be refactored into a "global" one that the Liens for particular commands can link to.

**Comment by wduquette** [comment]
Regarding Unicode, my plan with Molt so far has been:

* To use standard Molt strings with all due respect for their nature as arrays of bytes containing UTF-8 encoded characters that have to be extracted using the `chars()` iterator
* To presume that Rust's I/O code will properly handle text read from O/S files with respect to UTF-8, until I'm told differently
* To deal with anything further when it becomes relevant.

The tricky bit is that as a user of U.S. English and a guy who literally grew up with ASCII, I tend not to create files that contain anything other than standard ASCII, so I don't tend to run into the edge cases.

**Comment by wduquette** [comment]
Though I see you've added some Unicode test cases.  Excellent!

File: molt/src/commands.rs [hunk 1]
```diff
@@ -1121,6 +1121,31 @@ pub fn cmd_string_equal(_interp: &mut Interp, _: ContextID, argv: &[Value]) -> M
...
+    let start = if argv.len() == 5 {
+        argv[4].as_int()?
+    } else {
+        0
+    };
+
+    let pos = if start < 0 || start as usize >= haystack.len() {
+        -1
+    } else {
+        haystack[start as usize..]
```
Comment (wduquette) [inline molt/src/commands.rs]: I think there's a problem here.  If I'm not mistaken, you are starting the search at the nth byte, not the nth character.  There will be problems if `start` addresses the second or third byte of a multi-byte character; at the very least it won't get the right answer.  Unless you know something I don't?  Though of course it will work fine if all characters are one byte.

I think what you probably need to do is take `haystack.chars()`, skip the first `start` characters to create a new iterator, and then convert the head of the iterator back to a `&str` and do the find in that.

File: molt/src/commands.rs [hunk 2]
```diff
@@ -1150,6 +1150,39 @@ pub fn cmd_string_first(_interp: &mut Interp, _: ContextID, argv: &[Value]) -> M
...
+        } else {
+            Some(arg as usize)
+        }
+    } else {
+        Some(haystack.len() - 1)
+    };
+
+    let pos = match last {
+        Some(offset) =>
+            haystack[..=offset]
```
Comment (wduquette) [inline molt/src/commands.rs]: Same issue here as with `string first`; `start` is indexing into the bytes of the string, not the (possibly multi-byte) characters.

File: molt/src/commands.rs [hunk 3]
```diff
@@ -1189,6 +1189,76 @@ pub fn cmd_string_length(_interp: &mut Interp, _: ContextID, argv: &[Value]) ->
...
+        })
+        .filter(|(_, count, _)| *count > 0)
+        .collect::<Vec<_>>();
+    let matching_string = if nocase {
+        string.to_lowercase()
+    } else {
+        string.to_string()
+    };
+
+    let mut result = "".to_string();
```
Comment (wduquette) [inline molt/src/commands.rs]: Probably better as `let mut result = String::new();`

**Review by wduquette — CHANGES_REQUESTED** [review]
`string range` does the right thing, I think, to extract a substring given indices.  `string first` and `string last` need to be updated to use similar logic on the `haystack` when `start > 0` to avoid trouble.  Otherwise, it looks good.

**Comment by wduquette** [comment]
Haven't used the Github "code review" feature before; had I known it was going to display a big angry red "Changes requested" banner, I might have done it differently.

File: molt/src/commands.rs [hunk 4]
```diff
@@ -1121,6 +1121,31 @@ pub fn cmd_string_equal(_interp: &mut Interp, _: ContextID, argv: &[Value]) -> M
...
+    let needle = argv[2].as_str();
+    let haystack = argv[3].as_str();
+
+    let start = if argv.len() == 5 {
+        argv[4].as_int()?
+    } else {
+        0
+    };
+
+    let pos = if start < 0 || start as usize >= haystack.len() {
```
Comment (wduquette) [inline molt/src/commands.rs]: The length of a string is `haystack.chars().count()`.  Otherwise you're getting the number of UTF-8 bytes, which is different.  Apparently using `chars().count()` isn't flawless, either, as sometimes you get something that's only part of a printed character because the printed character is made up of multiple graphemes.  (I think grapheme is the right word.)

File: molt/src/commands.rs [hunk 5]
```diff
@@ -1150,6 +1150,39 @@ pub fn cmd_string_first(_interp: &mut Interp, _: ContextID, argv: &[Value]) -> M
...
+
+    let needle = argv[2].as_str();
+    let haystack = argv[3].as_str();
+
+    let last: Option<usize> = if argv.len() == 5 {
+        let arg = argv[4].as_int()?;
+
+        if arg < 0 {
+            None
+        } else if arg as usize >= haystack.len() {
```
Comment (wduquette) [inline molt/src/commands.rs]: As with `string first`, you need to be looking at `haystack.chars().count()`, not `haystack.len()`

File: molt/src/commands.rs (hunk 3 above)
Comment (dbohdan) [reply molt/src/commands.rs]: Done in 1c900a0.

File: molt/src/commands.rs (hunk 1 above)
Comment (dbohdan) [reply molt/src/commands.rs]: You are right.  Not only did I use the byte length and byte indices for `startIndex` and `lastIndex`, I also *returned* a byte index from `string first` and `string last`.  That is what `find()` and `rfind()` return.  I have rewritten `string first`/`last` to operate on chars and to return char indices.  I have also added tests that panic if the commands mishandle multibyte characters and fail if they return byte indices.  The Rust idiom for mapping a character index to a byte index seems to be `s.char_indices().nth(n)`.

File: molt/src/commands.rs (hunk 3 above)
Comment (dbohdan) [reply molt/src/commands.rs]: Done in 1c900a0.

File: molt/src/commands.rs (hunk 4 above)
Comment (dbohdan) [reply molt/src/commands.rs]: It would be nice and useful to have a command for splitting a string into graphemes.  I don't think graphemes should be the unit for indexing a string because other Unicode Tcl implementations work with code points.

**Comment by dbohdan** [comment]
>Haven't used the Github "code review" feature before; had I known it was going to display a big angry red "Changes requested" banner, I might have done it differently.

Oh, I didn't take it as angry.  (But I guess the red banner does look a little angry?)

I have resolved the change requests.  Please review.

**Comment by wduquette** [comment]
I'll take a look and most likely merge these later this morning.

**Comment by wduquette** [comment]
Re: Molt, string indices, graphemes, and grapheme.  Did a little looking around; per the Rust documentation, Rust characters are Unicode scalar values, which are similar to but not identical to code points:  "Unicode Scalar Value. Any Unicode code point except high-surrogate and low-surrogate code points."  The discussions of how to compute Rust string length that I'd read distinguished between characters and "grapheme clusters"; I took that to mean that a character was a single grapheme rather than a cluster of them, but now I see that the writer meant that a grapheme is a cluster of code points.

So characters in Rust strings are approximately code points.  I'm comfortable with that; my intent is that Molt should be an extension language for Rust, and so should adhere to Rust conventions.  Working with a string should have the same results at the TCL level as it does at the Rust level.  For example, Molt's notion of whitespace is the same as Rust's, and slightly different than Standard TCL's.

**Comment by dbohdan** [comment]
:tada: :tada: :tada:

**Comment by dbohdan** [comment]
>my intent is that Molt should be an extension language for Rust, and so should adhere to Rust conventions. Working with a string should have the same results at the TCL level as it does at the Rust level. For example, Molt's notion of whitespace is the same as Rust's, and slightly different than Standard TCL's.

I think this is the approach that makes the most sense for an embedded scripting language (that doesn't implement a particular standard).
